bpf/lib/nat.h

```diff
@@ -1460,13 +1778,26 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
 		return NAT_PUNT_TO_STACK;
 	};
 
-	if (snat_v6_rev_nat_can_skip(target, &tuple))
-		return NAT_PUNT_TO_STACK;
-	ret = snat_v6_rev_nat_handle_mapping(ctx, &tuple, &state, off, target);
-	if (ret > 0)
-		return CTX_ACT_OK;
-	if (ret < 0)
-		return ret;
+	if (icmp_rewrite_embedded) {
+		state = snat_v6_lookup(&tuple);
+		if (!state)
+			return NAT_PUNT_TO_STACK;
+		ret = snat_v6_icmp_rewrite_embedded(ctx, &tuple, state, off);
+		if (ret < 0)
+			return ret;
+		/* To rewrite headers for ingress we need the tuple to
+		 * use original next header.
+		 */
+		tuple.nexthdr = nexthdr;
```

Comment:
This logic can now move up, into the section where we set `icmp_rewrite_embedded = true`. Same for IPv4.